### PR TITLE
Ignore zero-weight samples during training

### LINKED ISSUE #143: Zero-weight minibatches poison PCD and UCD with NaNs
## Summary

Per-sample weights of zero should ignore those samples. However, if a minibatch contains only zero-weight samples, PCD first computes a normalized weighted mean by dividing by `mean(wd) == 0`. This produces `NaN`; multiplying the resulting gradient by the minibatch weight of zero does not recover it because `NaN * 0 == NaN`.

The overall weight vector can still have positive total weight, so this can occur simply because zero-weight samples are grouped in one minibatch.

Relevant code:

- [`wmean` normalization](https://github.com/cossio/RestrictedBoltzmannMachines.jl/blob/b1a7c57d/src/util/util.jl#L14-L32)
- [weighted positive phase before minibatch-weight correction](https://github.com/cossio/RestrictedBoltzmannMachines.jl/blob/b1a7c57d/src/train/pcd.jl#L75-L89)
- [the analogous UCD ordering](https://github.com/cossio/RestrictedBoltzmannMachines.jl/blob/b1a7c57d/src/train/ucd.jl#L201-L230)

## Reproduction

```julia
import RestrictedBoltzmannMachines as RBMs

data = Bool[
    0 1 0 1
    1 0 1 0
]
wts = [0.0, 0.0, 1.0, 1.0]

rbm = RBMs.BinaryRBM(
    zeros(2),
    zeros(1),
    fill(0.1, 2, 1),
)

RBMs.pcd!(
    rbm,
    data;
    wts,
    batchsize = 2,
    iters = 1,
    steps = 0,
    shuffle = false,
    rescale = false,
    zerosum = false,
)

@show all(isfinite, rbm.visible.par)
@show all(isfinite, rbm.hidden.par)
@show all(isfinite, rbm.w)
```

Observed at `b1a7c57d`:

```text
all(isfinite, rbm.visible.par) = true
all(isfinite, rbm.hidden.par) = false
all(isfinite, rbm.w) = false
```

The same ordering can affect centered and standardized PCD.

## Expected behavior

- Individual zero-weight samples contribute nothing to the objective and can be excluded when minibatches are formed.
- A minibatch with total weight zero contributes no training step and should be skipped before gradient, chain, optimizer, wrapper-statistics, and callback work.
- Skipping a zero-weight minibatch should not consume one of the requested `iters` parameter updates.
- If the complete training weight vector has total weight zero, reject it up front with a clear error because there is no data objective to estimate.

## Acceptance criteria

- The reproduction above leaves all parameters finite and still performs the requested number of updates using positive-weight samples.
- Mixed positive/zero-weight minibatches match the result obtained after removing the zero-weight samples.
- Globally all-zero weights raise a clear error before any model mutation.
- Plain, centered, and standardized PCD, plus UCD, use consistent semantics.

### PR DESCRIPTION
**Pull request by cossio**
## Summary

- validate weighted training inputs as finite, real, and nonnegative, with at least one positive weight
- remove zero-weight observations before moments, minibatches, chains, wrapper statistics, optimizer work, and callbacks
- make `iters` count completed parameter updates even when zero-weight observations would otherwise form empty batches
- defer default moments, fantasy particles, optimizer state, and UCD chain counts until after validation
- keep weighted training backend-safe on JLArrays/GPUArrays and compatible with Julia 1.10
- document the semantics and add focused regressions for plain, centered, and standardized PCD plus UCD

## Root cause

Weighted minibatch gradients normalized by the minibatch weight before applying the minibatch bias correction. A zero-total minibatch therefore divided by zero and produced NaNs, and the later multiplication by a zero batch factor could not recover them. Zero-weight observations also remained present in minibatch boundaries, wrapper statistics, and UCD chain initialization.

## Validation

- `julia --project=test test/zero_weight_training.jl`
- `julia --project=test test/pcd.jl`
- `julia --project=test test/ucd.jl`
- `julia --project=test test/centered.jl`
- `julia --project=test test/standardized.jl`
- `julia --project=test test/jlarrays.jl` (Julia 1.12 and Julia 1.10 with `allowscalar(false)`)
- `julia --project=. -e 'using Pkg; Pkg.test()'`
- `julia --project=docs docs/make.jl`

Fixes #143

### CODE REVIEW

**Comment by cossio** [comment]
Merged the latest `origin/master` normally in `cb6bb122` after PRs #144 and #147 landed. The only textual conflict was `CHANGELOG.md`; I retained the independent Unreleased entries for #140, #142, and #143. The overlapping CenteredRBM/StandardizedRBM and JLArrays changes merged cleanly, preserving both upstream zero-norm rescaling behavior and this PR's zero-weight training semantics.

Validated the reconciled tree with:
- `julia --project=test test/zero_weight_training.jl`
- `julia --project=test test/centered.jl`
- `julia --project=test test/standardized.jl`
- `julia --project=test test/jlarrays.jl`
- `julia --project=. -e 'using Pkg; Pkg.test()'`

All passed.

File: src/train/infinite_minibatches.jl [hunk 1]
```diff
@@ -60,3 +72,35 @@ function infinite_minibatches(
...
+    if npositive < length(wts)
+        # GPUArrays do not generally support logical indexing. Transfer only
+        # the mask used to build indices; indexing preserves the data backend.
+        positive_indices = findall(adapt(Array, positive))
+        data, wts = getobs(positive_indices, data, wts)
+    end
+
+    wts_mean = mean(wts)
+    isfinite(wts_mean) && wts_mean > 0 ||
+        throw(ArgumentError("the mean of the positive sample weights must be finite"))
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/train/infinite_minibatches.jl]: Fixed in ac008f35. Raw filtered weights are still passed unchanged to callbacks, while global and minibatch training weights are normalized by their respective maxima (using at least Float32) before moments, gradients, and wrapper-statistics work. The minibatch/global correction is reconstructed from the finite scales and normalized means, so finite weights such as [floatmax(Float64), floatmax(Float64)] no longer overflow and relative weighting is preserved. I added batchsize=1 scale-equivalence regressions for plain, centered, and standardized PCD plus UCD, a 70,000-sample Float16 accumulator regression, and JLArrays coverage with extreme weights. The focused CPU and JLArrays tests pass on Julia 1.10 and 1.12, and the full package suite including Aqua passes.

File: src/train/infinite_minibatches.jl [hunk 2]
```diff
@@ -60,3 +72,58 @@ function infinite_minibatches(
...
+        positive_indices = findall(adapt(Array, positive))
+        data, wts = getobs(positive_indices, data, wts)
+    end
+
+    # Weighted training is invariant to a common scale factor. Normalize before
+    # moments and gradients so finite weights cannot overflow their sum/mean.
+    # Keep the raw weights separately so callbacks continue to receive them.
+    scale = maximum(wts)
+    T = promote_type(typeof(scale), Float32)
+    training_wts = T.(wts) ./ T(scale)
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/train/infinite_minibatches.jl]: Fixed in 7a84e386. The normalization accumulator is now selected as promote_type(float(typeof(scale)), Float32): UInt128 and Int128 use Float64, BigInt uses BigFloat, and Float16 still gets the existing Float32 accumulation safeguard. This keeps the integral scale representable before division. I extended the scale-equivalence regressions with typemax(UInt128) for plain, centered, and standardized PCD plus UCD, including raw callback-weight checks. The focused suite passes on Julia 1.10 and 1.12, JLArrays passes on both, and the full package suite including Aqua passes.

File: src/train/infinite_minibatches.jl [hunk 3]
```diff
@@ -60,3 +72,58 @@ function infinite_minibatches(
...
+
+function _prepare_training_batch(
+    wts::AbstractVector,
+    normalization::NamedTuple,
+)
+    scale = maximum(wts)
+    T = promote_type(float(typeof(scale)), Float32)
+    training_wts = T.(wts) ./ T(scale)
+    batch_weight = (scale / normalization.scale) *
+        (mean(training_wts) / normalization.mean)
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/train/infinite_minibatches.jl]: Fixed in 4ec1f954. Both the minibatch scale and dataset scale are now converted to the common promoted accumulator type before division; the local type is also promoted with the global normalized-mean type so it cannot narrow relative to dataset preparation. The regression uses Float16[nextfloat(0), floatmax(Float16)] with batchsize 1 and verifies the small batch has a strictly positive, correctly scaled batch_weight. Focused tests pass on Julia 1.10/1.12, JLArrays passes on both, and the full suite including Aqua passes.

File: src/train/infinite_minibatches.jl [hunk 4]
```diff
@@ -60,3 +72,58 @@ function infinite_minibatches(
...
+        positive_indices = findall(adapt(Array, positive))
+        data, wts = getobs(positive_indices, data, wts)
+    end
+
+    # Weighted training is invariant to a common scale factor. Normalize before
+    # moments and gradients so finite weights cannot overflow their sum/mean.
+    # Keep the raw weights separately so callbacks continue to receive them.
+    scale = maximum(wts)
+    T = promote_type(float(typeof(scale)), Float32)
+    training_wts = T.(wts) ./ T(scale)
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/train/infinite_minibatches.jl:113]: Fixed in 4ec1f954. I added a shared accumulator selector that widens Float16 to Float32 and Float32 to Float64 while retaining Float64 for fixed-width integers and BigFloat for BigInt-derived scales. The regression uses the smallest positive Float32 subnormal, nextfloat(0), against floatmax(Float32), confirming the normalized small weight stays nonzero and its batch correction matches the widened-type ratio. Focused tests pass on Julia 1.10/1.12, JLArrays passes on both, and the full suite including Aqua passes.

File: src/train/infinite_minibatches.jl [hunk 5]
```diff
@@ -60,3 +72,58 @@ function infinite_minibatches(
...
+end
+
+function _prepare_training_batch(
+    wts::AbstractVector,
+    normalization::NamedTuple,
+)
+    scale = maximum(wts)
+    T = promote_type(float(typeof(scale)), Float32)
+    training_wts = T.(wts) ./ T(scale)
+    batch_weight = (scale / normalization.scale) *
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/train/infinite_minibatches.jl]: Fixed in 4ec1f954. I changed only the intended batch_weight calculation (while retaining scale, accumulator-type, and training-weight setup): it now computes T(scale) / T(normalization.scale), with T promoted against the global normalization type. A Float16[nextfloat(0), floatmax(Float16)] regression verifies the smallest positive batch retains a nonzero correction equal to the widened-type formula. Focused tests pass on Julia 1.10/1.12, JLArrays passes on both, and the full suite including Aqua passes.

File: src/train/infinite_minibatches.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (cossio) [reply src/train/infinite_minibatches.jl]: Fixed in 4ec1f954. The shared accumulator selector now widens Float16 to Float32 and Float32 to Float64, which preserves the ratio between any two positive finite values across each narrow IEEE type’s full range. The new Float32 regression is stricter than the reported case: it uses nextfloat(0) alongside floatmax(Float32) and verifies both the global normalized weight and the per-batch correction remain positive and numerically correct. Focused tests pass on Julia 1.10/1.12, JLArrays passes on both, and the full suite including Aqua passes.
